accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,493 @@
+
+- An extension only applies if the receiver type is a subtype of all `on` types.
+- An extension is more specific than another if for every `on` type in the latter, there is an `on` type in the former which is a proper subtype of that type, or the two are equivalent, and the former is a proper subtype of the latter when instantiated to bounds.
+- The trailing `?` makes the most sense if it is applied only once (it's the extension which accepts and understands `null` as a receiver), but for forwards compatibility, we will need to put it on every `on` type individually. All `on` types must be nullable in order to accept a nullable receiver.
+- There is no clear type to assign to `this` inside an instance extension method. For a mixin that's not a problem because it introduces a type by itself, and the combined super-interface is only used for `super` invocations. For extension, a statement like `var self = this;` needs to be assigned a useful type.
+
+The last item is the reason this feature is not something we will definitely do. We can start out without the feature and maybe add it later if it is necessary, but it's safer to start without it.
+
+### Extending Static Members
+
+The feature above only extends instance members. There is no way to add a new static member on an existing type, something that should logically be a *simpler* operation.
+
+We could allow
+
+```dart
+extension MyInt on num {
+  int get double => this * s;
+  static int get random => 4;
+}
+```
+
+to introduce both a `double` instance getter on `num` instances and a `random` getter on `num` itself, usable as `var n = num.random;`
+
+However, while this is possible, not all `on` types are class or mixin types. It is not clear what it would mean to put static methods on `on` types of extension like:
+
+- `extension X on Iterable<int>`
+- `extension X<T extends Comparable<T>> on Iterable<T>`
+- `extension X on int Function(int)`
+- `extension X on FutureOr<int>`
+- `extension X on int?`
+
+For the first two, we could put the static members on the `Iterable` class, but since the extension does not apply to *all* iterables, it is not clear that this is correct.
+
+For `int Function(int)` and `FutureOr<int>`, it's unclear how to call such a static method at all. We can denote`int Function(int)` with a type alias, but putting static members on type aliases is a new concept.  We could put the static method on `Function`, but that's not particularly discoverable, and why not require that they are put on `Functon` explicitly. For `FutureOr`, we could allow static members on `FutureOr` (which is a denotable type), but again it seems spurious. For `int?`, we could put the method on `int`, but why not  just require that it's on `int`.
+
+The issue here is that the type patterns used by `on` are much more powerful than what is necessary to put static members on class types.
+
+It would probably be more readable to introduce a proper static extension declaration:
```

Comment:
You could also simply disallow static members in extensions on non-simple class types.

That prohibits the perhaps useful cases of wanting a static member that can only be called through the extension itself, but I'm not sure that's that useful anyway.